lib/std/simd.zig

```diff
@@ -0,0 +1,369 @@
+// Note: Some functions here are known to malfunction on MIPS.
+// Functions whose tests fail on MIPS will trigger compile errors when compiled for that platform.
+
+const std = @import("std");
+const builtin = @import("builtin");
+
+pub const Vector = std.meta.Vector;
+
+/// Suggests a vector size for a given type and CPU, or null if it doesn't support SIMD natively.
+/// Not yet implemented for every CPU architecture.
+pub fn suggestVectorSizeForCpu(comptime T: type, cpu: std.Target.Cpu) ?usize {
+    switch (cpu.arch) {
+        .x86_64 => {
+            if (T == bool and std.Target.x86.featureSetHas(.prefer_mask_registers)) return 64;
+
+            const vector_bit_size = if (std.Target.x86.featureSetHas(.avx512f)) 512 else if (std.Target.x86.featureSetHas(.prefer_256_bit)) 256 else if (std.Target.x86.featureSetHas(.prefer_128_bit)) 128 else return null;
+            const element_bit_size = std.math.max(8, std.math.ceilPowerOfTwo(T, @bitSizeOf(T)));
+            return vector_bit_size / element_bit_size;
+        },
+        else => @compileError("No vector sizes for this CPU architecture have yet been recommended"),
+    }
+}
+
+/// Suggests a vector size for a given type, or null if the target CPU doesn't support SIMD natively.
+/// Not yet implemented for every CPU architecture.
+pub fn suggestVectorSize(comptime T: type) ?usize {
```

Comment:
This function looks redundant. Why is it needed?
If it is needed, only comment on the more likely used (shorter) function that wraps the longer one.

Reply:
I thought that needing a suggested vector size for the target platform would be the most common use-case, but that there might be other use-cases that I haven't foreseen. So `suggestVectorSize` is just a shortcut.

Reply:
I dont think `suggestVectorSize` is used that frequently to justify the additional function, which only adds `std.Target.Cpu`.
SIMD code will likely require to handle non-SIMD for portability, which requires usage of `std.Target`.
So I dont see the benefit.

Reply:
My rationale is that SIMD code will do either of these:

- Use the same vector sizes unconditionally. Zig lets you use any vector size you want, even if it's not native, much like with ints such as `u128`. This is simple to work with, and is fitting if you only need SIMD for a small, known amount of data that you'd expect to fit into one or two native vector registers. In this case, `suggestVectorSize` is not used at all.
- Adapt to the native vector size of the target platform, or avoid SIMD altogether if it's not native. `suggestVectorSize` will tell both if and how large vectors should be used, so I'm not sure why `std.Target` would be needed.

The reason I added both `suggestVectorSize` was to support the latter case, and I then decided to split it into `suggestVectorSizeForCpu` because I thought it'd be a waste if someone needed but couldn't get the suggested vector size for a different CPU. I'm not opposed to removing `suggestVectorSize` and having programmers use `suggestVectorSizeForCpu` directly. I simply thought that having it seemed more in style with the rest of stdlib, but I could be wrong.